### PR TITLE
API package

### PR DESCRIPTION
**Pull request by lukechampine**
siad is now just two files. The siad testing has moved to api. Functionality should be unaffected.

I consider the api package to be incomplete as of now. I'd like to make it more useful to siac, but I haven't yet hit upon the best way of accomplishing that.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+func (srv *Server) initAPI(addr string) {
+	mux := http.NewServeMux()
+
+	// Consensus API Calls
+	handleHTTPRequest(mux, "/consensus/status", srv.consensusStatusHandler)
+
+	// Server API Calls
+	handleHTTPRequest(mux, "/daemon/stop", srv.daemonStopHandler)
+	handleHTTPRequest(mux, "/daemon/update/apply", srv.daemonUpdateApplyHandler)
+	handleHTTPRequest(mux, "/daemon/update/check", srv.daemonUpdateCheckHandler)
```
Comment (DavidVorick) [inline api/api.go]: these should all be renamed to /server, and the function names should be srv.serverXXXHandler

File: api/server.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	return srv
+}
+
+func (srv *Server) updateWait() {
+	<-srv.minerUpdates
+	<-srv.walletUpdates
+}
+
+// TODO: move this to the state module?
+func (srv *Server) acceptBlock(conn modules.NetConn) error {
```
Comment (DavidVorick) [inline api/server.go]: do these functions even still get used?

File: api/server.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		return errors.New("block added, but it does not extend the state height")
+	}
+
+	srv.gateway.RelayBlock(b)
+	return nil
+}
+
+// TODO: move this to the tpool module?
+func (srv *Server) acceptTransaction(conn modules.NetConn) error {
+	var t consensus.Transaction
```
Comment (DavidVorick) [inline api/server.go]: same here

File: modules/host/host.go [hunk 4]
```diff
@@ -71,11 +71,11 @@ func New(state *consensus.State, tpool modules.TransactionPool, wallet modules.W
...
-			TotalStorage: 2e9,                                    // 2 GB
-			MaxFilesize:  300e6,                                  // 300 MB
-			MaxDuration:  5e3,                                    // Just over a month.
-			WindowSize:   288,                                    // 48 hours.
-			Price:        consensus.NewCurrency64(1000000000000), // 10^12
+			TotalStorage: 2e9,                           // 2 GB
+			MaxFilesize:  300e6,                         // 300 MB
+			MaxDuration:  5e3,                           // Just over a month.
+			WindowSize:   288,                           // 48 hours.
+			Price:        consensus.NewCurrency64(1e12), // 10^12
```
Comment (DavidVorick) [inline modules/host/host.go]: actually, before it is forgotten, can we change this to 100e6?

Comment (DavidVorick) [inline modules/host/host.go]: or maybe 1e9, both should be alright.

File: siad/daemon.go [hunk 5]
```diff
@@ -28,137 +25,49 @@ type DaemonConfig struct {
...
-	if err == consensus.ErrOrphan {
-		go d.gateway.Synchronize(conn.Addr())
-		return err
-	} else if err != nil {
-		return err
+	// bootstrap to the network
+	// TODO: probably a better way of doing this.
+	if !config.Siacore.NoBootstrap {
+		go gateway.Bootstrap(modules.BootstrapPeers[0])
 	}
```
Comment (DavidVorick) [inline siad/daemon.go]: should the bootstrapping be handled by the gateway? I feel like it's not really the daemon's job, it's more of the gateway's job.

The daemon is really just a tiny shell that strings everything together.

File: siad/main.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/mitchellh/go-homedir"
 	"github.com/spf13/cobra"
 
-	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/api"
```
Comment (DavidVorick) [inline siad/main.go]: I thought the api was going to be in the modules package?

**Comment by DavidVorick** [comment]
Some smaller comments.

I also thought the api was going to be in the modules folder.

File: api/server.go (hunk 2 above)
Comment (lukechampine) [inline api/server.go]: yes, they are registered to the gateway above.

File: siad/daemon.go (hunk 5 above)
Comment (lukechampine) [inline siad/daemon.go]: yeah. What should probably happen is that gateway.New should take a bootstrap flag. We also need a way to override the bootstrap peer. (For consistency, probably both a cmdline flag and a config value.)

**Comment by lukechampine** [comment]
I could see arguments for both sides. My first thought was, "well, none of the other modules import the API server." But that's also true of the host and hostdb. So it isn't 100% clear.

It seems like the modules folder is eating all the other packages, so I think down the line we may wind up moving all the modules stuff up a directory.

**Comment by DavidVorick** [comment]
Even if we move all the modules up a layer, we still need an explicit modules package.

**Comment by lukechampine** [comment]
Are you sure? Why is it required?

**Comment by DavidVorick** [comment]
The modules need to import eachother's interfaces without actually importing eachother. This prevents import cycles when doing testing. Otherwise the tpool would be importing a miner (which has a wallet) to help with testing, but the miner imports the tpool.

Using the modules package + interfaces, we can have testing scenarios where the tpool uses a miner but avoids import loops.

**Comment by lukechampine** [comment]
Sure, so all you do is move the interface definitions to a top-level file, or maybe to the types package.

**Pull request merged**